File: phase2_research/backend/app/schemas/whitebox.py
from pydantic import BaseModel, Field
from typing import Literal
from typing import Optional, Dict, Any

class WhiteboxRequest(BaseModel):
    fen: str = Field(..., description="FEN string of the current board position")
    engine: Literal["alphabeta", "mcts"] = Field(..., description="Type of engine: 'alphabeta' or 'mcts'")
    evaluator: Literal["material", "pst", "heuristic"] = Field("heuristic", description="Alpha-Beta evaluator to use")
    
    # Alpha-Beta specific hyperparameters
    depth: int = Field(
        3,
        ge=1,
        le=8,
        description="Search depth for Alpha-Beta. Web requests are capped at 8 to keep responses interactive.",
    )
    use_move_ordering: bool = Field(True, description="Enable heuristic move ordering (MVV-LVA)")
    
    # MCTS specific hyperparameters
    mcts_iterations: int = Field(
        100,
        ge=1,
        le=50000,
        description="Number of Monte Carlo iterations. Web requests are capped at 50000.",
    )
    mcts_exploration_constant: float = Field(
        1.414,
        gt=0.0,
        le=5.0,
        description="Exploration constant (c) in UCB1 formula, capped for the interactive Search Lab.",
    )

class Candidate(BaseModel):
    move: str
    evaluation: float = Field(
        ...,
        description="White-centric candidate evaluation: positive is better for White, negative is better for Black",
    )
    nodes: int = 0

class WhiteboxResponse(BaseModel):
    best_move: Optional[str] = Field(None, description="The best move selected by the engine")
    evaluation: float = Field(
        ...,
        description="White-centric evaluation: positive is better for White, negative is better for Black",
    )
    nodes_evaluated: int = Field(..., description="Total number of nodes evaluated/expanded")
    nps: int = Field(..., description="Nodes Per Second calculation")
    time_ms: int = Field(..., description="Total calculation time in milliseconds")
    instrumentation: Optional[Dict[str, Any]] = Field(None, description="Search instrumentation data")
    tree: Dict[str, Any] = Field(..., description="JSON serialized search tree for visualization")
    candidates: list[Candidate] = Field(default_factory=list)

File: phase2_research/frontend/src/types/whitebox.ts
export type WhiteboxEngine = 'alphabeta' | 'mcts';
export type EvaluatorName = 'material' | 'pst' | 'heuristic';

export interface WhiteboxRequest {
  fen: string;
  engine: WhiteboxEngine;
  depth: number;
  use_move_ordering: boolean;
  evaluator: EvaluatorName;
  mcts_iterations: number;
  mcts_exploration_constant: number;
}

export interface SearchTreeNodeMetadata {
  fen?: string;
  move_path?: string[];
  alpha?: number;
  beta?: number;
  depth_remaining?: number;
  reason?: string;
  visits?: number;
  wins?: number;
  ucb?: number;
  [key: string]: unknown;
}

export interface SearchTreeNode {
  id: string;
  name: string;
  value: number | null;
  node_type: string;
  is_pruned: boolean;
  metadata: SearchTreeNodeMetadata;
  children?: SearchTreeNode[];
}

export interface WhiteboxInstrumentation {
  evaluator_name?: EvaluatorName;
  nodes_visited?: number;
  leaf_evaluations?: number;
  leaf_nodes_evaluated?: number;
  cutoffs?: number;
  generated_children?: number;
  remaining_depth_counts?: Record<string, number>;
  children_by_remaining_depth?: Record<string, number>;
  branching_factor?: number;
  depth_node_counts?: Record<string, number>;
  [key: string]: unknown;
}

export interface Candidate {
  move: string;
  /** White-centric score: positive is better for White, negative is better for Black. */
  evaluation: number;
  nodes?: number;
}

export interface WhiteboxResult {
  best_move: string | null;
  /** White-centric score: positive is better for White, negative is better for Black. */
  evaluation: number;
  nodes_evaluated: number;
  nps: number;
  time_ms: number;
  instrumentation?: WhiteboxInstrumentation | null;
  candidates?: Candidate[];
  tree: SearchTreeNode;
}

export interface SearchConfig {
  engine: WhiteboxEngine;
  evaluator: EvaluatorName;
  depth: number;
  useMoveOrdering: boolean;
  mctsIterations: number;
  mctsExplorationConstant: number;
}
